Implements the ability to rename and restore internal functions

mathieuk opened the pull request:
Implemented by creating a HashTable to keep track of the original zend functions, and an userland function to walk that HashTable and restore the original functions.

mathieuk committed:
- Implements the ability to rename ZEND_INTERNAL_FUNCTION functions while keeping track of the original functions.
- Adds method restore_internal_functions() to restore all renamed functions to their original functions.
- Activated rename_internal_function.phpt, added restore_internal_functions.phpt

sebastianbergmann commented:
Johannes said that he will have to review these changes but he won't have time for this before April.

mathieuk commented:
April sure feels like a long time for small patch like this. Is there anything I can do to speed this along?

sebastianbergmann commented:
It is not as small as you think. And I really want Johannes to have a look at it. Sorry.

mathieuk commented:
Very well; it's not like I can't use it in the meantime :)

mathieuk commented:
Just a gentle reminder that we're currently in April, and I would still like to get it merged :)

johannes commented:
As you can see I made a few quick comments on the code. I didn't test it and didn't think through edge cases though.

Please mind that this for me is a small toy thing, in the priority somewhere behind vacation, friends&family, work, general PHP stuff, ...

Commit:
- Replaced space indent with tab / Comment fixes
- Use globals-> over THG() in php_test_helpers_init_globals
- Moved internals of PHP_FUNCTION(restore_internal_functions) to regular C function and are now calling that
- Resolve segfault on shutdown.

mathieuk commented:
@johannes: Thank you for taking the time to review my patch. I understand that this is a low priority thing for you. However, I also do think this adds an important feature to test_helpers for those stuck with ... special ... codebases :). That's why I'm just following through. I hope you guys didn't think my reminder was rude or anything.

I've just pushed another commit that should resolve all the issues you've mentioned, including the crash on shutdown ( i think, I'm not too sure about whether this _properly_ fixes it, as I'm not that fluent with C ).

Commit:
- Fix whitespace.

sebastianbergmann commented:
The test suite fails with your patch:

```
Build complete.
Don't forget to run 'make test'.

Segmentation fault
Segmentation fault

=====================================================================
PHP         : /usr/local/php-5.3/bin/php 
PHP_SAPI    : cli
PHP_VERSION : 5.3.7-dev
ZEND_VERSION: 2.3.0
PHP_OS      : Linux - Linux thinkpad 2.6.35-28-generic #49-Ubuntu SMP Tue Mar 1 14:39:03 UTC 2011 x86_64
INI actual  : /usr/local/src/test-helpers/tmp-php.ini
More .INIs  :  
CWD         : /usr/local/src/test-helpers
Extra dirs  : 
VALGRIND    : Not used
=====================================================================
TIME START 2011-04-11 13:38:39
=====================================================================
FAIL Loading test_helpers as zend_extension [tests/load_as_zend_extension.phpt] 
FAIL rename_function() and user-defined functions [tests/rename_function.phpt] 
FAIL rename_function() and internal functions [tests/rename_internal_function.phpt] 
FAIL restore_internal_functions() resets internal functions after renaming an internal function [tests/restore_internal_functions.phpt] 
FAIL set_exit_overload() [tests/set_exit_overload.phpt] 
FAIL debug_print_backtrace() should work inside the callback. [tests/set_new_overload_backtrace.phpt] 
FAIL A closure can be registered as a callback with set_new_overload() [tests/set_new_overload_closure.phpt] 
XFAIL Callback for the new operator is invoked for non-existing classes [tests/set_new_overload_error_non_existing_class.phpt]   XFAIL REASON: Currently not supported by the implementation but nice to have

FAIL A fatal error is triggered when the class returned by the callback does not exist [tests/set_new_overload_error_return_non_existing_class.phpt] 
FAIL A warning is triggered when the callback passed to set_new_overload() is not defined [tests/set_new_overload_error_undefined_callback.phpt] 
FAIL A function can be registered as a callback with set_new_overload() [tests/set_new_overload_function.phpt] 
FAIL A functor can be registered as a callback with set_new_overload() [tests/set_new_overload_functor.phpt] 
FAIL set_new_overload() can be called multiple times and unset_new_overload() works [tests/set_new_overload_multiple.phpt] 
FAIL A non-static method can be registered as a callback with set_new_overload() [tests/set_new_overload_non_static_method.phpt] 
FAIL A non-static method of the same object can be registered as a callback with set_new_overload() [tests/set_new_overload_non_static_method_same_object.phpt] 
FAIL A private method can be registered from the right context with set_new_overload() [tests/set_new_overload_private_method.phpt] 
FAIL A static method can be registered as a callback with set_new_overload() [tests/set_new_overload_static_method.phpt] 
FAIL set_new_overload() can be called after unset_new_overload() [tests/set_new_overload_unset_reset.phpt] 
FAIL set_exit_overload() [tests/unset_exit_overload.phpt] 
=====================================================================
TIME END 2011-04-11 13:38:40

=====================================================================
TEST RESULT SUMMARY
---------------------------------------------------------------------
Exts skipped    :    0
Exts tested     :   41
---------------------------------------------------------------------

Number of tests :   19                19
Tests skipped   :    0 (  0.0%) --------
Tests warned    :    0 (  0.0%) (  0.0%)
Tests failed    :   18 ( 94.7%) ( 94.7%)
Expected fail   :    1 (  5.3%) (  5.3%)
Tests passed    :    0 (  0.0%) (  0.0%)
---------------------------------------------------------------------
Time taken      :    1 seconds
=====================================================================

=====================================================================
FAILED TEST SUMMARY
---------------------------------------------------------------------
Loading test_helpers as zend_extension [tests/load_as_zend_extension.phpt]
rename_function() and user-defined functions [tests/rename_function.phpt]
rename_function() and internal functions [tests/rename_internal_function.phpt]
restore_internal_functions() resets internal functions after renaming an internal function [tests/restore_internal_functions.phpt]
set_exit_overload() [tests/set_exit_overload.phpt]
debug_print_backtrace() should work inside the callback. [tests/set_new_overload_backtrace.phpt]
A closure can be registered as a callback with set_new_overload() [tests/set_new_overload_closure.phpt]
A fatal error is triggered when the class returned by the callback does not exist [tests/set_new_overload_error_return_non_existing_class.phpt]
A warning is triggered when the callback passed to set_new_overload() is not defined [tests/set_new_overload_error_undefined_callback.phpt]
A function can be registered as a callback with set_new_overload() [tests/set_new_overload_function.phpt]
A functor can be registered as a callback with set_new_overload() [tests/set_new_overload_functor.phpt]
set_new_overload() can be called multiple times and unset_new_overload() works [tests/set_new_overload_multiple.phpt]
A non-static method can be registered as a callback with set_new_overload() [tests/set_new_overload_non_static_method.phpt]
A non-static method of the same object can be registered as a callback with set_new_overload() [tests/set_new_overload_non_static_method_same_object.phpt]
A private method can be registered from the right context with set_new_overload() [tests/set_new_overload_private_method.phpt]
A static method can be registered as a callback with set_new_overload() [tests/set_new_overload_static_method.phpt]
set_new_overload() can be called after unset_new_overload() [tests/set_new_overload_unset_reset.phpt]
set_exit_overload() [tests/unset_exit_overload.phpt]
=====================================================================

=====================================================================
EXPECTED FAILED TEST SUMMARY
---------------------------------------------------------------------
Callback for the new operator is invoked for non-existing classes [tests/set_new_overload_error_non_existing_class.phpt]  XFAIL REASON: Currently not supported by the implementation but nice to have

=====================================================================

You may have found a problem in PHP.
We would like to send this report automatically to the
PHP QA team, to give us a better understanding of how
the test cases are doing. If you don't want to send it
immediately, you can choose "s" to save the report to
a file that you can send us later.
```

johannes commented:
I fixed some of the issues, not all, in my 2ce16a7a9f261bebc16e402011edce02f29e8ff2

tests/restore_internal_functions.phpt still fails with

```
#0  0x08486a01 in zend_hash_get_current_data_ex (ht=0x8843078, pData=0x80465c4, pos=0x80465c8)
    at /home/johannes/src/php/src/php/php-src/branches/PHP_5_3/Zend/zend_hash.c:1166
1166            *pData = p->pData;
(gdb) bt
#0  0x08486a01 in zend_hash_get_current_data_ex (ht=0x8843078, pData=0x80465c4, pos=0x80465c8)
    at /home/johannes/src/php/src/php/php-src/branches/PHP_5_3/Zend/zend_hash.c:1166
#1  0xfe552aec in pth_restore_internal_functions () at /home/johannes/src/php-test-helpers/test_helpers.c:497
#2  0xfe552b1c in zif_restore_internal_functions (ht=0, return_value=0x88439d0, return_value_ptr=0x0, this_ptr=0x0, 
    return_value_used=0) at /home/johannes/src/php-test-helpers/test_helpers.c:544
#3  0x084a29dd in zend_do_fcall_common_helper_SPEC (execute_data=0x89a6f98)
    at /home/johannes/src/php/src/php/php-src/branches/PHP_5_3/Zend/zend_vm_execute.h:316
#4  0x084a6c4a in ZEND_DO_FCALL_SPEC_CONST_HANDLER (execute_data=0x89a6f98)
    at /home/johannes/src/php/src/php/php-src/branches/PHP_5_3/Zend/zend_vm_execute.h:1606
#5  0x084a1f2a in execute (op_array=0x88a2cc8)
    at /home/johannes/src/php/src/php/php-src/branches/PHP_5_3/Zend/zend_vm_execute.h:107
#6  0x0847914f in zend_execute_scripts (type=8, retval=0x0, file_count=3)
    at /home/johannes/src/php/src/php/php-src/branches/PHP_5_3/Zend/zend.c:1194
#7  0x0841724b in php_execute_script (primary_file=0x8047140)
    at /home/johannes/src/php/src/php/php-src/branches/PHP_5_3/main/main.c:2270
#8  0x085279e4 in main (argc=66, argv=0x80471c0)
    at /home/johannes/src/php/src/php/php-src/branches/PHP_5_3/sapi/cli/php_cli.c:1193
```

Where p in end_hash.c:1166 is efree'd memory being accessed.

johannes commented:
Oh, one more thing: While working on this I stumbled over that branch: 5f54aa2a9300e86af6764c8e4209e63f2bae00db, these functionalities are related so probably they should be merged once the issues here are resolved.

mathieuk commented:
Odd, I'm not getting any of these segfaults. Not the one Sebastian reported with my original change (tests ran fine here), but also not the one johannes reported with his fixes:

With Johannes changes merged and compiled in: 

<pre>
MatbookPro:php-test-helpers mathieu$ cat php.ini
zend_extension=.libs/test_helpers.so
date.timezone = Europe/Amsterdam
</pre>


<pre>
MatbookPro:php-test-helpers mathieu$ cat test.php
&lt;?php
function notdate() { echo "DATE", PHP_EOL; }
rename_function('date', 'date_old');
rename_function('notdate', 'date');
date();
restore_internal_functions();
</pre>


<pre>
MatbookPro:php-test-helpers mathieu$ php -v
PHP 5.3.5 (cli) (built: Mar  1 2011 22:21:34) 
Copyright (c) 1997-2010 The PHP Group
Zend Engine v2.3.0, Copyright (c) 1998-2010 Zend Technologies
</pre>


<pre>
MatbookPro:php-test-helpers mathieu$ php -c ./php.ini test.php
DATE
MatbookPro:php-test-helpers mathieu$
</pre>

johannes commented:
You're not using --enable-debug for you PHP build it seems. Maybe you could try that and if it still works try valgrind ...

```
USE_ZEND_ALLOC=0 valgrind --leak-check=full php -c ./php.ini test.php
```

Didn't check in detail, yet.

johannes commented:
The more I go through it, the more I think we should redo this based on Sara's apparoach in http://lxr.php.net/xref/PECL/runkit/runkit_functions.c#343 ... The example in b17f4cd1c2e738195d38f91e779bb0ec59d779d1 looks strange as the function tables look proper after the cleanup went through from what I saw in gdb but still the engine doesn't seem to free the function, which shouldn't be affected by anything at all.

mathieuk commented:
FYI, working on a different prototype by overloading INIT_FCALL_BY_NAME and DO_FCALL opcodes instead. This approach requires a lot less housekeeping and moving zend_function's around. Have it working for the simplest case (rename_function('a', 'b')), now trying to get it to work on <pre>rename_function('a', function() { echo 'b'; } );</pre> which is turning out to be kind of complicated as DO_FCALL only expects a simple function name as a string.

mathieuk commented:
Took a while, but I think I figured it out. Could you guys take a look at https://github.com/mathieuk/php-test-helpers/commit/cc1e4e0e4e3d2ae41faa0b30aa06d4913ed24422 and tell me what you think of this approach?

johannes commented:
The new approach means it won't work with situations where the function is called from somewhere inside PHP (like a callback) and adds a notable slowdown to each function call which can be bad in larger unit test suits.

mathieuk commented:
Good point on the internal callbacks. Pull request closed.
